### PR TITLE
[inductor] Add explicit ops.fma and use it in softmax_backward

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122915
* #115435
* #123269
* __->__ #122518

This allows us to generate an fma even when fp-fusion is disabled
in the compiler.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by Chillee** [comment]
Does fma help performance here?

**Comment by peterbell10** [comment]
I don't think it matters too much for performance, but in this case the fma improves precision. One of the test failures I was seeing in the next PR that disables fp fusion was fixed for me by reintroducing this fma. Specifically [`test_pattern_fails_with_tensor_factor_cuda`](https://github.com/pytorch/pytorch/actions/runs/8376550434/job/22938613473).

**Comment by Chillee** [comment]
That test specifically I also ran into a bunch of failures with - it's fundamentally extremely unstable. It's taking gradients of x/y wrt y, and the random values generated can be close to 0.

It's actually where I got the values from for the example that failed.

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -643,6 +650,13 @@ class OverridesData:
         triton="libdevice.erfcx({x})",
         name="special_erfcx",
     ),
+    fma=OverridesData(
+        type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.INT_TO_FLOAT,
+        cpp=lambda x, y, z: f"std::fma({x}, {y}, {z})",
```
Comment (lezcano) [inline torch/_inductor/codegen/common.py:655]: Can you append/prepend a PR changing all to this format?

File: torch/_inductor/decomposition.py [hunk 2]
```diff
@@ -79,6 +79,7 @@
 decomps_to_exclude = [
     aten._unsafe_index,
     aten._scaled_dot_product_flash_attention_for_cpu.default,  # See comments in torch/_decomp/decompositions.py
+    aten._softmax_backward_data,
```
Comment (lezcano) [inline torch/_inductor/decomposition.py:82]: why are you excluding it?

Comment (peterbell10) [reply torch/_inductor/decomposition.py:82]: If I don't exclude the core decomposition it would report a decomposition is already registered when it comes to register the inductor-specific decomp with the fma operation.

Comment (lezcano) [reply torch/_inductor/decomposition.py:82]: ah, I see, we already have a decomp for this one.

**Review by lezcano — APPROVED** [review]

**Review by eellison — COMMENTED** [review]
cc potential land race with https://github.com/pytorch/pytorch/pull/121313 @aorenste

**Comment by aorenste** [comment]
> cc potential land race with #121313 @aorenste

If you're ready to land go ahead - I still need to rebase to origin and let tests run again.

**Review by Chillee — APPROVED** [review]

File: torch/_inductor/codegen/common.py [hunk 3]
```diff
@@ -592,21 +592,28 @@ def func(x, y):
...
 class OverridesData:
     name: str
-    cpp: str
-    triton: Optional[str] = None  # None when not impl in libdevice/triton
-    cppvec: Optional[str] = None  # None when not impl in aten/.../vec
+    cpp: Union[str, Callable[..., str]]
+    # None when not impl in libdevice/triton
+    triton: Union[Optional[str], Callable[..., str]] = None
+    # None when not impl in aten/.../vec
+    cppvec: Union[Optional[str], Callable[..., str]] = None
```
Comment (jgong5) [inline torch/_inductor/codegen/common.py:616]: May I know why you chose to maintain codegen backend specific things in the common place instead of adding them to their respective subclasses as the original design does?

Comment (peterbell10) [reply torch/_inductor/codegen/common.py:616]: It has simplified adding a new `ops` method into a 7 line change that only touches a single file, which I think is pretty compelling. It does mean there are more places you need to look to find the overload which may be annoying, but it's still very grep-able so I don't think this is a major issue.

Also note that this just adds a method onto the `{Backend}Overrides` class in an unconventional way, rather than replacing the overrides mechanism.

**Comment by peterbell10** [comment]
Looks like this was merged in 9189d04cb1e0adeea4e4533280be2f0135637ea4, the bot just didn't close the PR. cc @huydhn
